### PR TITLE
Fix more issues with the gateway problem numbers.

### PR DESCRIPTION
**Pull request by drgrice1**
There were still cases where non-consecutively ordered gateway quizzes can cause errors.  In fact any time that a gateway quiz does not have a problem number 1 in the set definition, you will get an error.  In addition answer processing had issues that could cause errors. Basically, arrays storing problem data should generally not be accessed by the direct numerical index, but indirectly via the @probOrder array index.  There are some exceptions to this.  One exception is noted.

Note that this also fixes an issue with randomly ordered gateway quizzes.  The issue is perhaps minor, but the data stored in a column of the past_answers table was disjoint.  The past_answers_string and scores did not match up with the problem_id and source_file.  The source_file in that table is not used anywhere by webwork, but it still should match.  The problem_id is used.  In the past answers table, this mismatch leads to some rather unexpected results.  If a student takes multiple versions of the quiz, and you view the answers for all versions for that student, the answers clearly look like they come from different problems.  The same happens if you view answers for multiple students. One downside of this is that now the answers are not in the order that the problems appear for the student.  They are in the order they appear in the set definition.  That is consistent for all students and quiz versions though.

This fixes issue #1814 and the gateway random order/non-consecutive problem number issues ... hopefully all of them now.

### CODE REVIEW

File: lib/WeBWorK/ContentGenerator/GatewayQuiz.pm [hunk 1]
```diff
@@ -1663,11 +1666,11 @@ sub body {
 					# we assume that allowed is an even
 					#    number of minutes
 					my $allowed = ($set->due_date - $set->open_date)/60;
-					$scoreRecordedMessage[$i] = $r->maketext("Your score was not recorded because you have " .
-						"exceeded the time limit for this test. (Time taken: [_1] min; allowed: [_2] min.)",
+					$scoreRecordedMessage[ $probOrder[$i] ] = $r->maketext('Your score was not recorded because ' .
+						' you haveexceeded the time limit for this test. (Time taken: [_1] min; allowed: [_2] min.)',
```
Comment (pstaabp) [inline lib/WeBWorK/ContentGenerator/GatewayQuiz.pm:1670]: 'you haveexceeded the'  to 'you have exceeded the ...'

Comment (drgrice1) [reply lib/WeBWorK/ContentGenerator/GatewayQuiz.pm]: Good catch.  It is fixed.

**Review by pstaabp — APPROVED** [review]
I can reproduce the error and this seems to fix it.  I have a quiz with 1 problem per page and the problem numbers are listed sequentially without the skip in numbers.  For example, the quiz is 1,3,4,5 and the problems are listed as 1,2,3,4.  Personally, I think this is the right way to do this (student get confused when numbers are skipped).

**Comment by drgrice1** [comment]
That is actually the way it was before.  If the numbers were 1,3,4,5 in the set definition, then they are listed as 1,2,3,4 in the quiz itself.

To test this try creating a quiz with say 5 problems, and delete number 1.  If you do that with the current develop branch (or main) you will get an error when you try to start a quiz.  With this pull request you won't.

Another thing to look at is if you use randomization of problem order.  Create a quiz with say 5 problems each with a single text input answers and different source files.  Then open the quiz and enter the answers 1, 2, 3, 4, 5 into the problems in the order the problems appear.  Then look in the database at the past answers for the quiz.  You will see that the problem source file does not match the answers and scores.

**Review by drdrew42 — APPROVED** [review]
👍

**Pull request merged**